### PR TITLE
EWPP-1488: Adds access control handler for corporate entity types and sub entity types.

### PR DESCRIPTION
**Pull request by 22Alexandra**

### CODE REVIEW

**Review by nagyad — CHANGES_REQUESTED** [review]
The description of the ticket stated to create a single access control handler that will handle both cases. Let's adapt the code accordingly.

File: modules/oe_content_entity/src/CorporateEntityTypeAccessControlHandler.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected function checkAccess(EntityInterface $entity, $operation, AccountInterface $account) {
+    switch ($operation) {
+      // We grant access to the 'view label' operation to all users having
+      // permission to view published corporate entities.
+      case 'view label':
+        $corporate_entity_id = $entity->getEntityType()->getBundleOf();
+        return AccessResult::allowedIf($account->hasPermission("view published $corporate_entity_id") || $account->hasPermission('access content'));
```
Comment (nagyad) [inline modules/oe_content_entity/src/CorporateEntityTypeAccessControlHandler.php]: Try to adapt this to void calling ($account->hasPermission("view published $corporate_entity_id") when $corporate_entity_id is null.

**Review by nagyad — CHANGES_REQUESTED** [review]

**Review by nagyad — DISMISSED** [review]

File: modules/oe_content_sub_entity/tests/src/Kernel/SubEntityAccessControlHandlerTest.php [hunk 2]
```diff
@@ -276,6 +276,24 @@ protected function accessTestScenarios(): array {
...
+        'operation' => 'view label',
+        'expected_result' => AccessResult::neutral(),
+        'status' => 0,
+      ],
+      'user with view unpublished sub entities / view label / unpublished entity' => [
+        'permissions' => ['view unpublished sub entities'],
+        'operation' => 'view label',
+        'expected_result' => AccessResult::forbidden()->addCacheContexts(['user.permissions']),
+        'status' => 0,
+      ],
```
Comment (sergepavle) [inline modules/oe_content_sub_entity/tests/src/Kernel/SubEntityAccessControlHandlerTest.php]: As I understand these test cases are applied for access `view label` operation of the **sub-entity**, but according to changes it has to be applied to **sub-entity type** (bundle).

File: modules/oe_content_entity/tests/src/Kernel/CorporateEntityAccessControlHandlerTest.php [hunk 3]
```diff
@@ -189,6 +201,16 @@ protected function accessTestScenarios(): array {
...
+      'user with create, update, delete access / view label / published entity' => [
+        'permissions' => [
+          'create oe_corporate_entity_test test_bundle corporate entity',
+          'edit oe_corporate_entity_test test_bundle corporate entity',
+          'delete oe_corporate_entity_test test_bundle corporate entity',
+        ],
+        'operation' => 'view label',
+        'expected_result' => AccessResult::neutral()->addCacheContexts(['user.permissions'])->addCacheTags(['oe_corporate_entity_test:1']),
+        'status' => 1,
+      ],
```
Comment (sergepavle) [inline modules/oe_content_entity/tests/src/Kernel/CorporateEntityAccessControlHandlerTest.php]: As I understand these test cases are applied for access `view label` operation of the **corporate entity**, but according to changes it has to be applied to **corporate type entity** (bundle).

File: modules/oe_content_sub_entity/tests/src/Kernel/SubEntityAccessControlHandlerTest.php [hunk 4]
```diff
@@ -203,6 +212,21 @@ public function testAccess(): void {
...
+  public function testBundleAccess(): void {
+    $scenarios = $this->bundleAccessTestScenarios();
+    foreach ($scenarios as $scenario => $test_data) {
+      $user = $this->drupalCreateUser($test_data['permissions']);
+      $this->assertAccessResult(
+        $test_data['expected_result'],
+        $this->bundleAccessControlHandler->access($this->subEntity, $test_data['operation'], $user, TRUE),
+        sprintf('Failed asserting access for "%s" scenario.', $scenario)
+      );
+    }
```
Comment (sergepavle) [inline modules/oe_content_sub_entity/tests/src/Kernel/SubEntityAccessControlHandlerTest.php]:
```suggestion
    $scenarios = $this->bundleAccessTestScenarios();
    $type_storage = $this->container->get('entity_type.manager')->getStorage('oe_sub_entity_type_test');
    $subentity_bundle = $type_storage->create([
      'id' => 'test_bundle',
      'label' => 'Test bundle',
    ]);
    $subentity_bundle->save();
    foreach ($scenarios as $scenario => $test_data) {
      $user = $this->drupalCreateUser($test_data['permissions']);
      $this->assertAccessResult(
        $test_data['expected_result'],
        $this->bundleAccessControlHandler->access($subentity_bundle, $test_data['operation'], $user, TRUE),
        sprintf('Failed asserting access for "%s" scenario.', $scenario)
      );
    }
```

**Review by sergepavle — DISMISSED** [review]

File: modules/oe_content_entity/src/CorporateEntityTypeAccessControlHandler.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      // We grant access to the 'view label' operation to all users having
+      // permission to view published content.
+      case 'view label':
+        $corporate_entity_id = $entity->getEntityType()->getBundleOf();
+        if ($corporate_entity_id) {
+          return AccessResult::allowedIf($account->hasPermission("view published $corporate_entity_id") || $account->hasPermission('access content'))->cachePerPermissions();
+        }
+        else {
+          return AccessResult::allowedIfHasPermission($account, 'access content');
+        }
```
Comment (nagyad) [inline modules/oe_content_entity/src/CorporateEntityTypeAccessControlHandler.php]: We don't have to wrap this into an else.

File: modules/oe_content_entity/tests/src/Kernel/CorporateEntityAccessControlHandlerTest.php [hunk 6]
```diff
@@ -30,6 +30,20 @@ class CorporateEntityAccessControlHandlerTest extends EntityKernelTestBase {
...
+   * @var \Drupal\oe_content_entity\CorporateEntityTypeAccessControlHandler
+   */
+  protected $bundleAccessControlHandler;
+
+  /**
+   * A corporate entity to be used in tests.
+   *
+   * @var \Drupal\oe_content_entity\Entity\CorporateEntityTypeInterface
+   */
+  protected $entity;
```
Comment (nagyad) [inline modules/oe_content_entity/tests/src/Kernel/CorporateEntityAccessControlHandlerTest.php]: I don't see why we need this global variable at all. You create an entity during setup, that is being used only in one test case so far, for other test cases it's not being used but still it is being created.

**Review by nagyad — CHANGES_REQUESTED** [review]

**Review by nagyad — DISMISSED** [review]

**Review by sergepavle — DISMISSED** [review]

**Review by sergepavle — APPROVED** [review]

**Review by upchuk — APPROVED** [review]
Approving on behalf of Adam.

**Pull request merged**